lib/elixir/lib/kernel.ex

```diff
@@ -4117,6 +4117,64 @@ defmodule Kernel do
   end
 
   @doc """
+  Generates a macro suitable for use in guard expressions.
+
+  It raises at compile time if the definition uses expressions that aren't
+  allowed in guards, and otherwise creates a macro that can be used both inside
+  or outside guards.
+
+  ## Example
+
+      defmodule Integer.Guards do
+        defguard is_even(value) when is_integer(value) and rem(value, 2) == 0
+      end
+
+      defmodule Integer.Utils do
+        import Integer.Guards
+
+        def is_even_and_large?(value) when is_even(value) and value > 100, do: true
+        def is_even_and_large?(_), do: false
+
+        def is_even_and_small?(value) do
+          if is_even(value) and value <= 100, do: true, else: false
+        end
```

Comment:
Not the best examples, they can simplified to

```elixir
def is_even_and_large?(value), do: is_even(value) and value > 100
def is_even_and_small?(value), do: is_even(value) and value <= 100
```

A better example (https://en.wikipedia.org/wiki/Collatz_conjecture):

```elixir
@doc """
The Collatz conjecture is a conjecture in mathematics named after Lothar Collatz. It can be
summarized as follows. Take any positive integer n. If n is even, divide it by 2 to get n / 2. If n
is odd, multiply it by 3 and add 1 to obtain 3n + 1. Repeat the process indefinitely. The conjecture
is that no matter what number you start with, you will always eventually reach 1.

This function takes a positive integer number and returns the number of steps it takes to reach 1.
"""
def collatz(n) when n > 0, do: collatz(n, 0)

defp collatz(1, step_count), do: step_count
defp collatz(n, step_count) when is_even(n), do: collatz(div(n, 2), step_count + 1)
defp collatz(n, step_count), do: collatz(3*n + 1, step_count + 1)
```